From 561c803491a0ce7a2442e6ab8f8dc7d05367b85d Mon Sep 17 00:00:00 2001
From: zyexro <zyexro@proton.me>
Date: Tue, 21 Apr 2026 21:57:48 +0100
Subject: [PATCH] selinux/avc: add slow avc audit hook v2

Signed-off-by: zyexro <zyexro@proton.me>
---
 security/selinux/avc.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/security/selinux/avc.c b/security/selinux/avc.c
index 0ca02231a848..d66f436f53e4 100644
--- a/security/selinux/avc.c
+++ b/security/selinux/avc.c
@@ -768,6 +768,10 @@ static void avc_audit_post_callback(struct audit_buffer *ab, void *a)
 	}
 }
 
+#ifdef CONFIG_KSU
+extern int ksu_handle_slow_avc_audit_new(u32 tsid, u16 *tclass);
+#endif
+
 /* This is the slow part of avc audit with big stack footprint */
 noinline int slow_avc_audit(struct selinux_state *state,
 			    u32 ssid, u32 tsid, u16 tclass,
@@ -782,6 +786,11 @@ noinline int slow_avc_audit(struct selinux_state *state,
 		return 0;
 	if (!denied)
 		return 0;
+#ifdef CONFIG_KSU
+	ksu_handle_slow_avc_audit_new(tsid, &tclass);
+	if (!tclass)
+		return 0;
+#endif
 
 	if (!a) {
 		a = &stack_data;
-- 
2.53.0

